File: internal/adapter/mongodb/prize_repository.go
package mongodb

import (
	"context"
	"errors"
	"spahtmx/internal/domain"

	"log/slog"

	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo"
	"go.mongodb.org/mongo-driver/v2/mongo/options"
)

type PrizeMongoRepository struct {
	DB *mongo.Database
}

type PrizeMongo struct {
	ID                bson.ObjectID   `bson:"_id"`
	Year              string          `bson:"year"`
	Category          string          `bson:"category"`
	OverallMotivation string          `bson:"overallMotivation,omitempty"`
	Laureates         []LaureateMongo `bson:"laureates"`
}

type LaureateMongo struct {
	Firstname  string `bson:"firstname"`
	Surname    string `bson:"surname"`
	Motivation string `bson:"motivation"`
	Share      string `bson:"share"`
}

func ToPrizeDomain(p PrizeMongo) domain.Prize {

	return domain.Prize{
		Year:              p.Year,
		Category:          p.Category,
		OverallMotivation: p.OverallMotivation,
		Laureates: func() []domain.Laureate {
			var laureates []domain.Laureate
			for _, l := range p.Laureates {
				laureates = append(laureates, domain.Laureate{
					Firstname:  l.Firstname,
					Surname:    l.Surname,
					Motivation: l.Motivation,
					Share:      l.Share,
				})
			}
			return laureates
		}(),
	}

}

func FromPrizeDomain(prize domain.Prize) (*PrizeMongo, error) {

	var uid bson.ObjectID
	var err error

	if prize.ID == "" {
		uid = bson.NewObjectID()
	} else {
		uid, err = bson.ObjectIDFromHex(prize.ID)
		if err != nil {
			return nil, err
		}
	}

	return &PrizeMongo{
		ID:                uid,
		Year:              prize.Year,
		Category:          prize.Category,
		OverallMotivation: prize.OverallMotivation,
		Laureates: func() []LaureateMongo {
			var laureates []LaureateMongo
			for _, l := range prize.Laureates {
				laureates = append(laureates, LaureateMongo{
					Firstname:  l.Firstname,
					Surname:    l.Surname,
					Motivation: l.Motivation,
					Share:      l.Share,
				})
			}
			return laureates
		}(),
	}, nil
}

func (m PrizeMongoRepository) GetPrizes(ctx context.Context) ([]domain.Prize, error) {
	cursor, err := m.DB.Collection("prize").Find(ctx, bson.D{}, options.Find().SetLimit(10))
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)

	var u []PrizeMongo
	err = cursor.All(ctx, &u)
	if err != nil {
		return nil, err
	}

	var domainPrizes []domain.Prize
	for _, prize := range u {
		domainPrizes = append(domainPrizes, ToPrizeDomain(prize))
	}
	return domainPrizes, nil
}

func (m PrizeMongoRepository) GetPrize(ctx context.Context, id string) (domain.Prize, error) {
	objid, err := bson.ObjectIDFromHex(id)
	if err != nil {
		return domain.Prize{}, domain.ErrInvalidInput
	}
	user := m.DB.Collection("prize").FindOne(ctx, bson.D{{Key: "_id", Value: objid}}, nil)
	var p PrizeMongo
	err = user.Decode(&p)
	if err != nil {
		if errors.Is(err, mongo.ErrNoDocuments) {
			return domain.Prize{}, domain.PrizeNotFound
		}
		slog.Error("Database error in GetPrize", "error", err, "id", id)
		return domain.Prize{}, domain.ErrInternal
	}
	return ToPrizeDomain(p), nil
}

func (m PrizeMongoRepository) InsertPrizes(ctx context.Context, prizes []domain.Prize) error {
	//TODO implement me
	panic("implement me")
}

File: src/adapter/mongodb/prize_repository.rs
use async_trait::async_trait;
use mongodb::{Database, bson::{doc, oid::ObjectId}};
use futures_util::stream::StreamExt;
use crate::domain::models::{Prize, Laureate};
use crate::domain::repositories::PrizeRepository;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct PrizeMongo {
    #[serde(rename = "_id")]
    pub id: ObjectId,
    pub year: String,
    pub category: String,
    #[serde(rename = "overallMotivation", skip_serializing_if = "Option::is_none")]
    pub overall_motivation: Option<String>,
    pub laureates: Vec<LaureateMongo>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct LaureateMongo {
    pub firstname: Option<String>,
    pub surname: Option<String>,
    pub motivation: Option<String>,
    pub share: Option<String>,
}

impl From<PrizeMongo> for Prize {
    fn from(p: PrizeMongo) -> Self {
        Prize {
            id: p.id.to_hex(),
            year: p.year,
            category: p.category,
            overall_motivation: p.overall_motivation,
            laureates: Some(p.laureates.into_iter().map(Into::into).collect()),
        }
    }
}

impl From<LaureateMongo> for Laureate {
    fn from(l: LaureateMongo) -> Self {
        Laureate {
            firstname: l.firstname,
            surname: l.surname,
            motivation: l.motivation,
            share: l.share,
        }
    }
}

pub struct PrizeMongoRepository {
    pub db: Database,
}

#[async_trait]
impl PrizeRepository for PrizeMongoRepository {
    async fn get_prizes(&self) -> Result<Vec<Prize>, Box<dyn std::error::Error + Send + Sync>> {
        let collection = self.db.collection::<PrizeMongo>("prize");
        let mut cursor = collection.find(doc! {}).await?;
        let mut prizes = Vec::new();
        while let Some(prize) = cursor.next().await {
            prizes.push(prize?.into());
        }
        Ok(prizes)
    }

    async fn get_prize(&self, id: &str) -> Result<Prize, Box<dyn std::error::Error + Send + Sync>> {
        let obj_id = ObjectId::parse_str(id)?;
        let collection = self.db.collection::<PrizeMongo>("prize");
        let prize = collection.find_one(doc! { "_id": obj_id }).await?;
        prize.map(Into::into).ok_or_else(|| "Prize not found".into())
    }

    async fn get_prizes_by_year(&self, year: &str) -> Result<Vec<Prize>, Box<dyn std::error::Error + Send + Sync>> {
        let collection = self.db.collection::<PrizeMongo>("prize");
        let mut cursor = collection.find(doc! { "year": year }).await?;
        let mut prizes = Vec::new();
        while let Some(prize) = cursor.next().await {
            prizes.push(prize?.into());
        }
        Ok(prizes)
    }

    async fn get_prizes_by_category(&self, category: &str) -> Result<Vec<Prize>, Box<dyn std::error::Error + Send + Sync>> {
        let collection = self.db.collection::<PrizeMongo>("prize");
        let mut cursor = collection.find(doc! { "category": category }).await?;
        let mut prizes = Vec::new();
        while let Some(prize) = cursor.next().await {
            prizes.push(prize?.into());
        }
        Ok(prizes)
    }

    async fn get_prizes_by_category_and_year(&self, category: &str, year: &str) -> Result<Vec<Prize>, Box<dyn std::error::Error + Send + Sync>> {
        let collection = self.db.collection::<PrizeMongo>("prize");
        let mut cursor = collection.find(doc! { "category": category, "year": year }).await?;
        let mut prizes = Vec::new();
        while let Some(prize) = cursor.next().await {
            prizes.push(prize?.into());
        }
        Ok(prizes)
    }

    async fn get_categories(&self) -> Result<Vec<String>, Box<dyn std::error::Error + Send + Sync>> {
        let collection = self.db.collection::<PrizeMongo>("prize");
        let categories = collection.distinct("category", doc! {}).await?;
        Ok(categories.into_iter().map(|b| b.as_str().unwrap_or_default().to_string()).collect())
    }

    async fn get_years(&self) -> Result<Vec<String>, Box<dyn std::error::Error + Send + Sync>> {
        let collection = self.db.collection::<PrizeMongo>("prize");
        let years = collection.distinct("year", doc! {}).await?;
        Ok(years.into_iter().map(|b| b.as_str().unwrap_or_default().to_string()).collect())
    }
}
